### PR TITLE
Add typedefinitions

### PR DESCRIPTION
**Pull request by championswimmer**
Signed-off-by: Arnav Gupta <arnav@codingblocks.com>

### CODE REVIEW

**Comment by chriskrycho** [comment]
Possibly useful for comparison: [here are mine as a Gist](https://gist.github.com/chriskrycho/9ed39076ccb698023e57e59080fedecd).

**Comment by machty** [comment]
I have no TypeScript experience to understand what all of this means/does or how to compare @championswimmer's and @chriskrycho's solutions; could you (and other one else) explain/compare/contrast?

**Comment by chriskrycho** [comment]
I'll take a look later this week and compare. Cc. also @dwickern and @mike-north, both of whom have looked at this before.

**Comment by championswimmer** [comment]
@chriskrycho 
I think I need to change my `is[Status]` flags to `ComputedProperty<boolean>` otherwise, my definitions look like they're more exhaustive.

**Comment by dwickern** [comment]
The typings @chriskrycho also supports typed arguments to `.perform()` instead of `...any[]`

**Comment by mike-north** [comment]
@machty and co. - at risk of stating the obvious, misaligned type information runs the risk of providing a poor dev experience (even worse than the complete absence of type information).

At present, anything involving generator functions is a bit tricky due to gaps in TS ([this](https://github.com/Microsoft/TypeScript/issues/2983) is the main thing we're waiting on -- as I understand it the underlying type system can handle this already, it's just a debate over the right syntax)

Even after iterating on this, we may want to cut a pre-release and get a few people consuming this before considering a stable release.

File: types/Task.d.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    readonly isFinished: boolean;
+
+    /** True if the task is still running. */
+    readonly isRunning: boolean;
+
+    /** True if the task instance is fulfilled. */
+    readonly isSuccessful: boolean;
+
+    /** Describes the state that the task instance is in. Can be used for debugging, or potentially driving some UI state. */
+    readonly state: 'dropped' | 'cancelled' | 'finished' | 'running' | 'waiting';
```
[the remark this answers is not in the document]
Comment (championswimmer) [reply types/Task.d.ts]: Sounds good !

File: types/Task.d.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+     * and the reason for the cancelation, e.g. "TaskInstance
+     * 'doStuff' was canceled because the object it lives on
+     * was destroyed or unrendered"
+     */
+    debug(): void;
+    drop(): void;
+    enqueue(): void;
+    group(groupPath: any): any;
+    keepLatest(): void;
+    maxConcurrency(n: number): void;
```
[the remark this answers is not in the document]
Comment (championswimmer) [reply types/Task.d.ts]: Yeah correct, just saw that

File: types/Task.d.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    readonly value: any;
+
+    /** Cancels the task instance. Has no effect if the task instance has already been canceled or has already finished running. */
+    cancel(): void;
+
+    catch(): RSVP.Promise<any>;
+
+    finally(): RSVP.Promise<any>;
+
+    then(): RSVP.Promise<any>;
```
[the remark this answers is not in the document]
Comment (championswimmer) [reply types/Task.d.ts]: There is a type for Promise.ThenFunction and Promise.CatchFunction. will use those

**Comment by championswimmer** [comment]
@mike-north @chriskrycho @machty 
Can you please review this once again

**Comment by chriskrycho** [comment]
I'll take a look later; and in particular I will drop it in in our app and make sure it covers all the cases my own typings linked above cover, and if not, suggest how to make sure it does!

Thanks for working hard on this!

**Comment by championswimmer** [comment]
Any updates 🙈

**Comment by chriskrycho** [comment]
Yes, sorry; ended up very swamped this week.

I've taken *some* time with this and will block out an hour to dig in on Monday. I *can* say from reading it carefully that we do need to pull in some of the things my typings have, because they give you completion and type checking for `perform` invocations and a few other things – but I can *also* say that your work especially on having docs will be *invaluable* as we work to get everything tightened down and merged!

**Comment by chriskrycho** [comment]
@championswimmer I'm super sorry this keeps getting delayed. We're trying to ship some stuff at work and have hit roadblock after roadblock basically since I got back from EmberConf. 😭 I have a handful of things I'm blocking out time for today, and this is one of them. I'll probably open a PR against your PR to hit the couple pieces I see are missing. You've got a fair number of things I missed, and vice versa!

**Comment by chriskrycho** [comment]
And we ran into *more* derailing today. I *will* get to this; I apologize for the many delays.

**Comment by championswimmer** [comment]
Haha, no problem @chriskrycho :)

**Comment by perlun** [comment]
Ping @chriskrycho, we will all be _incredibly_ grateful once this lands. 😇

**Comment by machty** [comment]
(i have no idea what I'm talking about) is this something that we can just merge and tweak/improve over time? or is there a lot of damage if we publish something not 100% perfect?

**Comment by championswimmer** [comment]
@chriskrycho got any time this week 🙈

**Comment by chriskrycho** [comment]
😭 maybe. Hopefully, even!

**Comment by josemarluedke** [comment]
Any news on this? I would love to get this as well.

**Comment by chriskrycho** [comment]
As my friends back in North Carloina might have said: “Lord wiling and the creek don’t rise…” I’ll get to this on Tuesday.

**Comment by championswimmer** [comment]
@chriskrycho 
If you could point me towards those types that can improve, I can spend some time on this, this week and get this done with

**Comment by mike-north** [comment]
### TLDR

I feel pretty strongly that a decorator(s) should be considered as **part of an MVP** for the first stable ember-concurrency release that contains type information. 

---

I've just spent about a week focusing on Ember+TS ergonomics, and type definitions for ember-concurrency have become one of the "sore thumbs" that stick out (only because so much excellent work has been put into other areas).

Currently (with types much like those in this PR) the way you use e-c with TypeScript is like is like

```ts
// This pattern is equivalent to typical use of ember-concurrency
export default class XFoo extends Component.extend({
   // stuff that MUST go on the prototype goes up here
   doSearch: task(function*(this: XFoo) {
      const title = this.get('title'); 
       ...
   }) 
}) {
  public title = '';
  public description = '';
  // vast majority of other interesting stuff goes down here.
}
```
One challenge (trap?) teams are facing is that there's a "cleaner looking" pattern that's 100% wrong, and newcomers to ES6 classes and TypeScript seem to want to use it 
```ts
//// THE PATTERN BELOW IS WRONG AND SHOULD NOT BE USED ////

// TS (or ES6 w/ class fields) version
export default class XFoo extends Component {
  public title = '';
  public description = '';
  public doSearch = task(function*(this: XFoo) {
      const title = this.get('title'); 
       ...
  });
}

// ES6 version
export default class XFoo extends Component {
  constructor() {
     super();
     this.doSearch = task(function*(this: XFoo) {
       const title = this.get('title'); 
        ...
      });
  }
}
```
At its root, there are several factors that allow teams to fall into this trap, including 
* The temptation to incorrectly think of `class extends Component { /* stuff */}` as equivalent to `Component.extend({ /* stuff */}`
* The widespread lack of understanding of how prototypal inheritance works, in general
* Ember's productivity-increasing "guardrails" result in developers not **needing** to deeply understand JavaScript objects in order to build an app

I feel pretty strongly that a decorator(s) should be considered as **part of an MVP** for the first stable ember-concurrency release that contains type information. Consumers don't have to use it, they should be there, and the recommended method of consumption. The goal of this decision would be to guide developers toward ensuring their e-c tasks are defined on prototypes and not each instance.

Decorator(s) could live in a separate addon (i.e., live in ember-decorators), but the release(s) should be coordinated such that ES6/TS consumers are encouraged to use them (or at least nearly 100% of consumers in this group should know the approach is available and less error-prone).

Example usage might look like
```ts
import Component from '@ember/component' 
import { task, restartable, maxConcurrency } from 'ember-decorators/task';

export default class XFoo extends Component {
  public title = '';
  public description = '';

  @maxConcurrency(3)
  @restartable
  @task
  public doSearch*() {
      const title = this.get('title'); 
       ...
  }
}
```

**Comment by championswimmer** [comment]
@mike-north 
This is so very true.
90% of all newcomers to Ember will **most definitely** treat `extend Component` similar to `Component.extend({})`, and they would have no clue what is going wrong.

**Comment by chriskrycho** [comment]
Unfortunately, the scope issue is what led me to experiment with class fields in the first place. 😞 There’s no workaround for it in the traditional Ember Object world that gets you the relevant types without running into exactly what you found here. The only other thing I’ve thought of (but haven’t tried) is defining the generator function outside the class, including its this argument, and binding it in the prototype. I suspect that will just show the same type error, though.

**Comment by gossi** [comment]
Let's warm up this topic. Today, we have decorators for tasks 🎉

Also I did created my own typings, too. Because I find existing ones just afterwards, they are here: https://github.com/gossi/spomoda/blob/master/client/types/ember-concurrency/index.d.ts

What makes me curious is the way to import a task.

- You did: `import { Task } from 'ember-concurrency';`
- While I was going with: `import Task from 'ember-concurrency';` (felt natural to me)

As this might be another aspect (since the code still looks totally different) and might have an impact on it (when it may be will refactored ... to TS?).

**Comment by mfeckie** [comment]
Awaiting this PR with excitement.  As it is so far, I don't think any the proposed solutions cover encapsulated tasks.

For example

```ts
foo = task({ 
  count: 1, 
  *perform() {
    this.incrementProperty('count');
    // do other stuff
  } 
})
```

**Comment by buschtoens** [comment]
I just had an idea and would love your opinions: https://github.com/machty/ember-concurrency-decorators/issues/30

**Comment by buschtoens** [comment]
I opened #319 to have another go at this.

**Comment by maxfierke** [comment]
Thanks so much for your work on this! However, we're closing this in favor of #357, which has been merged and will be part of 1.2.0
